### PR TITLE
run optimizations on pre-profiled graph

### PR DESCRIPTION
**Pull request by Krovatkin**
This is the first stab at running profile-insensitive optimizations on pre-profiled graphs. Running those optimizations has a potential to simplify graphs greatly before GuardElimination and GuardElimination should be able to remove more guards.

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary and remediations
As of commit 0f32570:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0f325706e6d9f79ab3522a955c01d0fd720cdc31).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3990818) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76156860) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76156860))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/validate-docker-version.py 
Auto-merging .circleci/validate-docker-version.py 
CONFLICT (add/add): Merge conflict in .circleci/scripts/should_run_job.sh 
Auto-merging .circleci/scripts/should_run_job.sh 
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -67,32 +83,34 @@ ProfilingGraphExecutorImpl::ProfilingGraphExecutorImpl(
 ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(Stack& stack) {
   std::lock_guard<std::mutex> lock(compile_mutex);
   GRAPH_DEBUG("Running ProfilingGraphExecutorImpl ", this);
+
```
Comment (zdevito) [inline torch/csrc/jit/profiling_graph_executor_impl.cpp]: Can we separate getPlanFor into one function that does the mutable changes to the executors state, and others that are pure functions that run various optimization phase? I am having trouble following what things need to get set, and how the flags interact with this.

**Review by zdevito — APPROVED** [review]
Looks good. I just have  few organization comments.

File: torch/csrc/jit/profiling_graph_executor_impl.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: fixed

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 2]
```diff
@@ -57,6 +71,8 @@ std::shared_ptr<Graph> ProfilingGraphExecutorImpl::prepareGraph(
     const std::shared_ptr<Graph>& graph,
     Stack& stack) {
   auto g = graph->copy();
+  GRAPH_DUMP("prepareGraph:", g);
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: fixed

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 3]
```diff
@@ -35,6 +35,20 @@ std::atomic<bool>& getExecutorMode() {
...
+    std::shared_ptr<Graph>& copy) {
+  if (!getGraphExecutorOptimize()) {
+    runRequiredPasses(copy);
+    return;
+  }
+
+  LowerGradOf(*copy);
+  runRequiredPasses(copy);
+  ConstantPropagation(copy);
+  runOptimization(copy, /*unroll=*/false);
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: fixed

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 4]
```diff
@@ -35,6 +35,20 @@ std::atomic<bool>& getExecutorMode() {
...
 }
 
+void ProfilingGraphExecutorImpl::runProfilingInsensitiveOptimizations(
+    std::shared_ptr<Graph>& copy) {
+  if (!getGraphExecutorOptimize()) {
+    runRequiredPasses(copy);
+    return;
+  }
+
+  LowerGradOf(*copy);
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: fixed

**Pull request closed without merging**